Ginger/GingerCore/ALM/OctaneCore.cs

```diff
@@ -37,12 +44,147 @@ public class OctaneCore : ALMCore
+            catch (Exception ex)
+            {
+                Reporter.ToLog(eLogLevel.ERROR, "Failed to get Test Lab with REST API", ex);
+            }
+            return testlabPathList;
+        }
+
+        private string GetRootFolderId()
+        {
+            return Task.Run(() =>
+            {
+                LogicalQueryPhrase f1 = new LogicalQueryPhrase("name", "Application Modules", ComparisonOperator.Equal);
+                LogicalQueryPhrase f2 = new LogicalQueryPhrase("path", "000000", ComparisonOperator.Equal);
+                return octaneRepository.GetEntities<ApplicationModule>(GetLoginDTO(), new List<IQueryPhrase>() { f1, f2 });
+            }).Result.FirstOrDefault().Id.ToString();
+        }
+
+        private string GetTestLabFolderId(string separateAti, string separateAtIMinusOne)
+        {
+            if (!ExploredApplicationModule.ContainsKey(separateAti))
+            {
+                LogicalQueryPhrase parent = new LogicalQueryPhrase("id", separateAtIMinusOne, ComparisonOperator.Equal);
+                CrossQueryPhrase qdParent = new CrossQueryPhrase("parent", parent);
+                LogicalQueryPhrase lp = new LogicalQueryPhrase("name", separateAti, ComparisonOperator.Equal);
+                List<IQueryPhrase> filter = new List<IQueryPhrase>() { qdParent, lp };
+                List<ApplicationModule> listnodes = Task.Run(() =>
+                {
+                    return octaneRepository.GetEntities<ApplicationModule>(GetLoginDTO(), filter);
+                }).Result;
+
+                ExploredApplicationModule.Add(listnodes.FirstOrDefault().Name, listnodes.FirstOrDefault().Id);
+                return listnodes.FirstOrDefault().Id;
+            }
+            else
+            {
+                return ExploredApplicationModule[separateAti];
+            }
+        }
+
+        private LoginDTO GetLoginDTO()
+        {
+            if (this.loginDto == null)
+            {
+                AlmResponseWithData<AlmDomainColl> domains = Task.Run(() =>
+                {
+                    return octaneRepository.GetLoginProjects(ALMCore.DefaultAlmConfig.ALMUserName, ALMCore.DefaultAlmConfig.ALMPassword, ALMCore.DefaultAlmConfig.ALMServerURL);
+                }).Result;
+                AlmDomain domain = domains.DataResult.Where(f => f.DomainName.Equals(ALMCore.DefaultAlmConfig.ALMDomain)).FirstOrDefault();
+                ProjectArea project = domain.Projects.Where(p => p.ProjectName.Equals(ALMCore.DefaultAlmConfig.ALMProjectName)).FirstOrDefault();
+                this.loginDto = new LoginDTO()
+                {
+                    User = ALMCore.DefaultAlmConfig.ALMUserName,
+                    Password = ALMCore.DefaultAlmConfig.ALMPassword,
+                    Server = ALMCore.DefaultAlmConfig.ALMServerURL,
+                    SharedSpaceId = domain.DomainId,
+                    WorkSpaceId = project.ProjectId
+                };
+            }
+            return this.loginDto; ;
```

Comment:
Codacy found an issue: [Remove this empty statement.](https://app.codacy.com/manual/Ginger-Automation/Ginger/pullRequest?prid=6317725)